include/mqtt/endpoint.hpp

```diff
@@ -7431,11 +7441,22 @@ class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, Loc
         auto fixed_header = static_cast<std::uint8_t>(*b);
         switch (get_control_packet_type(fixed_header)) {
         case control_packet_type::publish: {
-            auto sp = std::make_shared<std::string>(b, e);
-            restore_serialized_message(basic_publish_message<PacketIdBytes>(sp->begin(), sp->end()), sp);
+            auto size = static_cast<std::size_t>(std::distance(b, e));
+            shared_ptr_array spa { new char[size] };
```

Comment:
In C++20, you can do instead:

`std::make_shared<char[]>(size);`

Which avoids a secondary allocation behind the scenes.

For now, using boost::make_shared<char[]>(size) will avoid that problem.

Perhaps introduce a new CMake option?

MQTT_USE_BOOST_SHARED_PTR

If it's enabled, then mqtt_cpp uses boost::shared_ptr, and can do this with one allocation instead of two.

If it's disabled, then do an #if cpp20 check, and if compiling with cpp 20, use std::make_shared<char[]>(size);

And lastly, use the current code if neither of the above are true.

Reply:
I think that the following code is good enough:

```cpp
#if !defined(MQTT_SHARED_PTR_ARRAY_HPP)
#define MQTT_SHARED_PTR_ARRAY_HPP

#ifdef MQTT_STD_SHARED_PTR_ARRAY

#include <memory>

namespace mqtt {

using shared_ptr_array = std::shared_ptr<char []>;
using shared_ptr_const_array = std::shared_ptr<char const []>;

inline shared_ptr_array make_shared_ptr_array(std::size_t size) {
#if __cplusplus > 201703L // C++20 date is not determined yet
    return std::make_shared<char[]>(size);
#else  // __cplusplus > 201703L
    return new char[size];
#endif // __cplusplus > 201703L
}

} // namespace mqtt

#else  // MQTT_STD_SHARED_PTR_ARRAY

#include <boost/shared_ptr.hpp>
#include <boost/smart_ptr/make_shared.hpp>

namespace mqtt {

using shared_ptr_array = boost::shared_ptr<char []>;
using shared_ptr_const_array = boost::shared_ptr<char const []>;

inline shared_ptr_array make_shared_ptr_array(std::size_t size) {
    return boost::make_shared<char[]>(size);
}

} // namespace mqtt

#endif // MQTT_STD_SHARED_PTR_ARRAY

#endif // MQTT_SHARED_PTR_ARRAY_HPP
```

Reply:
Implemented. Use boost::shared_ptr and boost::make_shared by default.